src/openfl/_internal/renderer/opengl/batcher/BatchRenderer.hx

```diff
@@ -0,0 +1,354 @@
+package openfl._internal.renderer.opengl.batcher;
+
+import haxe.ds.Vector;
+
+import lime.utils.Float32Array;
+import lime.utils.UInt16Array;
+import lime.graphics.GLRenderContext;
+import lime.graphics.opengl.GLBuffer;
+import openfl.display.BlendMode;
+import openfl._internal.renderer.opengl.GLBlendModeManager;
+import openfl._internal.renderer.opengl.GLShaderManager;
+import openfl._internal.renderer.opengl.batcher.BitHacks.*;
+import openfl._internal.renderer.opengl.batcher.Quad;
+
+#if gl_stats
+import openfl._internal.renderer.opengl.stats.GLStats;
+import openfl._internal.renderer.opengl.stats.DrawCallContext;
+#end
+
+// inspired by pixi.js SpriteRenderer
+class BatchRenderer {
+	var gl:GLRenderContext;
+	var blendModeManager:GLBlendModeManager;
+	var shaderManager:GLShaderManager;
+	var maxQuads:Int;
+	var maxTextures:Int;
+
+	var shader:MultiTextureShader;
+	var indexBuffer:GLBuffer;
+	var vertexBuffer:GLBuffer;
+	var vertexBufferDatas:Array<Float32Array>;
+
+	var renderedQuadCount:Int;
+	var renderedQuads:Array<Quad>;
+	var groups:Vector<RenderGroup>;
+	var boundTextures:Vector<TextureData>;
+	var tick = 0;
+	var textureTick = 0;
+
+	public var projectionMatrix:Float32Array;
+
+	public function new(gl:GLRenderContext, blendModeManager:GLBlendModeManager, shaderManager:GLShaderManager, maxQuads:Int) {
+		this.gl = gl;
+		this.blendModeManager = blendModeManager;
+		this.shaderManager = shaderManager;
+		this.maxQuads = maxQuads;
+
+		// determine amount of textures we can draw at once and generate a shader for that
+		maxTextures = gl.getParameter(gl.MAX_TEXTURE_IMAGE_UNITS);
+		shader = new MultiTextureShader(gl, maxTextures);
+
+		// a singleton vector we use to track texture binding when rendering
+		boundTextures = new Vector(maxTextures);
+
+		// preallocate blocks of memory for vertex buffers of different sizes
+		// smallest one can render just one quad, biggest one can render maximum amount
+		vertexBufferDatas = [];
+		var i = 1, l = nextPow2(maxQuads);
+		while (i <= l) {
+			vertexBufferDatas.push(new Float32Array(i * 4 * MultiTextureShader.floatPerVertex));
```

Comment:
I'm not sure if this makes sense actually. I ported that from PIXI code, but I'm now thinking that just having a single `Float32Array` of maximum size and updating it with `bufferSubData` would be cleaner and better. Need to test the performance though, so let's leave this for later.